FlexibleDiff/SectionedChangeset.swift

```diff
@@ -82,57 +82,63 @@ public struct SectionedChangeset {
+		let allInsertions = metadata.inserts.union(IndexSet(moveDestinationLookup.values))
+		let allRemovals = metadata.removals.union(IndexSet(moveDestinationLookup.keys))
 
 		mutatedSections = Array()
 		mutatedSections.reserveCapacity(Int(current.count))
 
 		var moves: [Changeset.Move] = []
 		var mutations = IndexSet()
 
-		for (offset, section) in current.enumerated() where !metadata.inserts.contains(offset) {
-			let predeletionOffset: Int
+		for (sourceOffset, section) in previous.enumerated() where !metadata.removals.contains(sourceOffset) {
+			let destinationOffset: Int
 			let isMove: Bool
 
-			if let moveSource = moveSourceLookup[offset] {
-				predeletionOffset = moveSource
+			if let moveDestination = moveDestinationLookup[sourceOffset] {
+				destinationOffset = moveDestination
 				isMove = true
 			} else {
-				let preinsertionOffset = offset - allInsertions.count(in: 0 ..< offset)
-				predeletionOffset = preinsertionOffset + allRemovals.count(in: 0 ... preinsertionOffset)
+				let postRemovalOffset = sourceOffset - allRemovals.count(in: 0 ..< sourceOffset)
+
+				var insertionsBeforeOffset = IndexSet()
+				for insertion in allInsertions where insertion <= postRemovalOffset + insertionsBeforeOffset.count {
+					insertionsBeforeOffset.insert(insertion)
+				}
+
+				destinationOffset = postRemovalOffset + insertionsBeforeOffset.count
 				isMove = false
 			}
 
-			let previousIndex = previous.index(previous.startIndex, offsetBy: numericCast(predeletionOffset))
-			let previousItems = items(previous[previousIndex])
-			let currentItems = items(section)
+			let previousItems = items(section)
+			let currentIndex = current.index(current.startIndex, offsetBy: numericCast(destinationOffset))
+			let currentItems = items(current[currentIndex])
 
 			let changeset = Changeset(previous: previousItems,
 			                          current: currentItems,
 			                          identifier: itemIdentifier,
 			                          areEqual: areItemsEqual)
 
 			let isMutated = !changeset.hasNoChanges
-				|| metadata.mutations.contains(predeletionOffset)
-				|| mutatedMoveDests.contains(offset)
+				|| metadata.mutations.contains(sourceOffset)
+				|| mutatedMoveDests.contains(destinationOffset)
 
 			if isMutated {
-				let section = SectionedChangeset.MutatedSection(source: predeletionOffset,
-				                                                destination: offset,
-				                                                changeset: changeset)
+				let section = SectionedChangeset.MutatedSection(source: sourceOffset,
+																destination: destinationOffset,
+																changeset: changeset)
 				mutatedSections.append(section)
 			}
 
 			if isMove {
-				moves.append(Changeset.Move(source: predeletionOffset,
-				                            destination: offset,
-				                            isMutated: isMutated))
+				moves.append(Changeset.Move(source: sourceOffset,
+											destination: destinationOffset,
```

Comment:
[Nit] I guess one of the tab vs space argument is that Xcode sucks with handling tabs...